src/stream/src/executor/backfill/cdc/state.rs

```diff
@@ -95,37 +220,46 @@ impl<S: StateStore> SingleTableState<S> {
         self.source_state_handler.init_epoch(epoch)
     }
 
-    pub async fn check_finished(&self) -> StreamExecutorResult<bool> {
+    pub async fn restore_state(&self) -> StreamExecutorResult<CdcStateRecord> {
         let mut key = self.split_id.to_string();
         key.push_str(BACKFILL_STATE_KEY_SUFFIX);
-        match self.source_state_handler.get(key.into()).await? {
+        let is_finished = match self.source_state_handler.get(key.into()).await? {
             Some(row) => match row.datum_at(1) {
-                Some(ScalarRefImpl::Jsonb(jsonb_ref)) => Ok(jsonb_ref.as_bool()?),
+                Some(ScalarRefImpl::Jsonb(jsonb_ref)) => jsonb_ref.as_bool()?,
                 _ => unreachable!("invalid backfill persistent state"),
             },
-            None => Ok(false),
-        }
+            None => false,
+        };
+        Ok(CdcStateRecord {
+            is_finished,
+            ..Default::default()
+        })
     }
 
-    /// When snapshot read stream ends, we should persist two states:
-    /// 1) a backfill finish flag to denote the backfill has done
-    /// 2) a consumed binlog offset to denote the last binlog offset
-    /// which will be committed to the state store upon next barrier.
-    pub async fn mutate_state(
-        &mut self,
-        last_binlog_offset: Option<CdcOffset>,
-    ) -> StreamExecutorResult<()> {
+    pub async fn mutate_state(&mut self, state_item: &CdcStateRecord) -> StreamExecutorResult<()> {
+        // skip if unfinished for single backfill
+        if !state_item.is_finished {
+            return Ok(());
+        }
+
+        // When single backfill is finished, we should persist two states:
+        // 1) a finish flag to denote the backfill has done
+        // 2) a consumed binlog offset to denote the last binlog offset
+        // which will be committed to the state store upon next barrier.
         let mut key = self.split_id.to_string();
         key.push_str(BACKFILL_STATE_KEY_SUFFIX);
         // write backfill finished flag
         self.source_state_handler
-            .set(key.into(), JsonbVal::from(Value::Bool(true)))
+            .set(
+                key.into(),
```

Comment:
Since for the backfill for single table scenario, we didn't introduce a new operator to the query plan. Just wraps the source executor in the cdc backfill executor and reuse the state table of `Source`. We access source state table only via point query with the ke column, so it won't break.

Reply:
True. If only get is used in the streaming job, there won't be unexpected result. But when querying source state via SQL, we can see something like this:
```
partition_id  |                                                                                                                                                                      offset_info                                                                                                                                                                       
---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1002          | {"split_info": {"mysql_split": {"inner": {"snapshot_done": false, "split_id": 1002, "start_offset": "{\"sourcePartition\":{\"server\":\"RW_CDC_1002\"},\"sourceOffset\":{\"transaction_id\":null,\"ts_sec\":1699866101,\"file\":\"binlog.000012\",\"pos\":157,\"server_id\":1},\"isHeartbeat\":true}"}}, "pg_split": null}, "split_type": "mysql-cdc"}
 1002_backfill | true
(2 rows)
```

This looks very strange to me. How about storing the backfill finished flag inside the json as a new field under "mysql_split"? This issue is introduced in #12535 and not related to the changes in this PR. I am okay with handling it separately.

Reply:
Yeah, it is a bit dirty. As mentioned in https://github.com/risingwavelabs/risingwave/pull/13276#discussion_r1390589039, I plan to refactor the plan of  single table cdc backfill to make the cdc backfill become a separate operator instead of just wraps the source inside it.